sdk/python/feast/pyspark/launcher.py

```diff
@@ -0,0 +1,145 @@
+import os
+import pathlib
+from typing import TYPE_CHECKING, List, Union
+
+from feast.config import Config
+from feast.constants import (
+    CONFIG_SPARK_DATAPROC_CLUSTER_NAME,
+    CONFIG_SPARK_DATAPROC_PROJECT,
+    CONFIG_SPARK_DATAPROC_REGION,
+    CONFIG_SPARK_DATAPROC_STAGING_LOCATION,
+    CONFIG_SPARK_LAUNCHER,
+    CONFIG_SPARK_STANDALONE_MASTER,
+)
+from feast.data_source import BigQuerySource, DataSource, FileSource
+from feast.feature_table import FeatureTable
+from feast.pyspark.abc import JobLauncher, RetrievalJob
+from feast.value_type import ValueType
+
+if TYPE_CHECKING:
+    from feast.client import Client
+
+
+def _standalone_launcher(config: Config) -> JobLauncher:
+    from feast.pyspark.launchers import standalone
+
+    return standalone.StandaloneClusterLauncher(
+        config.get(CONFIG_SPARK_STANDALONE_MASTER)
+    )
+
+
+def _dataproc_launcher(config: Config) -> JobLauncher:
+    from feast.pyspark.launchers import gcloud
+
+    return gcloud.DataprocClusterLauncher(
+        config.get(CONFIG_SPARK_DATAPROC_CLUSTER_NAME),
+        config.get(CONFIG_SPARK_DATAPROC_STAGING_LOCATION),
+        config.get(CONFIG_SPARK_DATAPROC_REGION),
+        config.get(CONFIG_SPARK_DATAPROC_PROJECT),
+    )
+
+
+_launchers = {"standalone": _standalone_launcher, "dataproc": _dataproc_launcher}
+
+
+def resolve_launcher(config: Config) -> JobLauncher:
+    return _launchers[config.get(CONFIG_SPARK_LAUNCHER)](config)
+
+
+_SOURCES = {
+    FileSource: ("file", "file_options", {"path": "file_url", "format": "file_format"}),
+    BigQuerySource: ("bq", "bigquery_options", {"table_ref": "table_ref"}),
+}
+
+
+def source_to_argument(source: DataSource):
+    common_properties = {
+        "field_mapping": dict(source.field_mapping),
+        "event_timestamp_column": source.event_timestamp_column,
+        "created_timestamp_column": source.created_timestamp_column,
+        "date_partition_column": source.date_partition_column,
+    }
+
+    kind, option_field, extra_properties = _SOURCES[type(source)]
+
+    properties = {
+        **common_properties,
+        **{
+            k: getattr(getattr(source, option_field), ref)
+            for k, ref in extra_properties.items()
+        },
+    }
+
+    return {kind: properties}
+
+
+def feature_table_to_argument(client: "Client", feature_table: FeatureTable):
+    return {
+        "features": [
+            {"name": f.name, "type": ValueType(f.dtype).name}
+            for f in feature_table.features
+        ],
+        "project": "default",
+        "name": feature_table.name,
+        "entities": [
+            {"name": n, "type": client.get_entity(n).value_type}
+            for n in feature_table.entities
+        ],
+        "max_age": feature_table.max_age.ToSeconds() if feature_table.max_age else None,
+    }
+
+
+def start_historical_feature_retrieval_spark_session(
```

Comment:
I'm not 100% follow what is this for? I'd expect this to be within the local launcher as it assumes that you have Spark env setup.

Reply:
This function is useful for cases where, the user has their own SparkSession, in a jupyter notebook for example. In such case, it might be more convenient to return a Spark dataframe, rather than the actual output.